demos/link-prediction/random-walks/ctdne-link-prediction.ipynb

```diff
@@ -0,0 +1,680 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "StellarGraph: Undirected multigraph\n",
+      " Nodes: 151, Edges: 37929\n",
+      "\n",
+      " Node types:\n",
+      "  default: [151]\n",
+      "    Features: none\n",
+      "    Edge types: default-default->default\n",
+      "\n",
+      " Edge types:\n",
+      "    default-default->default: [37929]\n",
+      "\n",
+      "Training examples: 9524 positive links, 9524 negative links\n",
+      "Validation examples: 3119 positive links, 3119 negative links\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(graph.info())\n",
+    "print()\n",
+    "print(f\"Training examples: {len(pos)} positive links, {len(neg)} negative links\")\n",
+    "print(\n",
+    "    f\"Validation examples: {len(pos_test)} positive links, {len(neg_test)} negative links\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Running random walks\n",
+    "\n",
+    "We try to keep the setup comparable between the use of temporal and biased (static) random walks. For temporal walks, the input parameter is defined in terms of the total number of context windows you are interested in obtaining, which differs from the traditional approach of specifying the number of walks to run per node in the graph. We calculate the number of context windows we need in terms of the traditional parameters as:\n",
+    "```\n",
+    "num_cw = num_nodes * num_walks_per_node * (walk_length - cw_size)\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of temporal random walks: 1763\n",
+      "Number of static random walks: 1510\n"
+     ]
+    }
+   ],
+   "source": [
+    "from stellargraph.data import TemporalRandomWalk, BiasedRandomWalk\n",
+    "\n",
+    "num_walks_per_node = 10\n",
+    "walk_length = 80\n",
+    "cw_size = 10\n",
+    "num_cw = (\n",
+    "    len(graph.nodes()) * num_walks_per_node * (walk_length - cw_size)\n",
+    ")  # as recommended in reference paper\n",
+    "\n",
+    "temporal_rw = TemporalRandomWalk(graph)\n",
+    "temporal_walks = temporal_rw.run(\n",
+    "    num_cw=num_cw,\n",
+    "    cw_size=cw_size,\n",
+    "    max_walk_length=walk_length,\n",
+    "    initial_edge_bias=None,\n",
+    "    walk_bias=\"exponential\",\n",
+    "    p_walk_success_threshold=0.2,\n",
+    ")\n",
+    "\n",
+    "static_rw = BiasedRandomWalk(graph)\n",
+    "static_walks = static_rw.run(\n",
+    "    nodes=graph.nodes(), n=num_walks_per_node, length=walk_length\n",
+    ")\n",
+    "\n",
+    "\n",
+    "print(\"Number of temporal random walks: {}\".format(len(temporal_walks)))\n",
+    "print(\"Number of static random walks: {}\".format(len(static_walks)))"
+   ]
```

Comment:
How long does the walking in this cell take? Is it worth splitting it up into a `TemporalRandomWalk` cell followed by a `BiasedRandomWalk` one so that changing the parameters for one doesn't require re-running the other one too?

Reply:
Yeah they do take a little while to run (maybe a few minutes?) I'll split them